diff --git a/browser/base/content/csshtmltree.js b/browser/base/content/csshtmltree.js
--- a/browser/base/content/csshtmltree.js
+++ b/browser/base/content/csshtmltree.js
@@ -101,16 +101,28 @@ CssHtmlTree.prototype = {
       this.createStyleGroupViews();
 
       this.viewedDocument = this.viewedElement.ownerDocument;
       this._template(this.templateHeader, this.header, this);
       this._template(this.templateGroups, this.groups, this);
       this._template(this.templateFooter, this.footer, this);
 
       this.specificityInput.checked = this.showSpecificity;
+
+      // Update the web page to display the selected source filter.
+      let sheetList = this.styleDocument.getElementById("sheetList");
+      let sheetItem = null;
+      let sourceFilter = this.cssLogic.sourceFilter;
+      for (let i = 0, n = sheetList.itemCount; i < n; i++) {
+        sheetItem = sheetList.getItemAtIndex(i);
+        if (sheetItem && sheetItem.value == sourceFilter) {
+          sheetList.selectedIndex = i;
+          break;
+        }
+      }
     } else {
       this.viewedDocument = null;
       this.header.innerHTML = "";
       this.groups.innerHTML = "";
       this.footer.innerHTML = "";
     }
   },
 
@@ -145,16 +157,32 @@ CssHtmlTree.prototype = {
     aEvent.preventDefault();
     if (aEvent.target && aEvent.target.pathElement &&
         aEvent.target.pathElement != InspectorUI.selection) {
       InspectorUI.inspectNode(aEvent.target.pathElement);
     }
   },
 
   /**
+   * The oncommand event handler for the sheets menulist.
+   * @param {Event} aEvent the DOM event.
+   */
+  sheetChange: function CssHtmlTree_sheetChange(aEvent) {
+    let target = aEvent.target;
+    if (target.value == this.cssLogic.sourceFilter) {
+      return;
+    }
+
+    this.cssLogic.sourceFilter = target.value;
+
+    this.createStyleGroupViews();
+    this._template(this.templateGroups, this.groups, this);
+  },
+
+  /**
    * Provide access to the path to get from document.body to the selected
    * element
    */
   get pathElements()
   {
     return CssLogic.getShortNamePath(this.viewedElement);
   },
 
@@ -476,16 +504,18 @@ PropertyView.prototype = {
     // TODO: We need a 'RuleView' to contain the view functionality from the
     // rule template. The implementation of this has been started, but it is
     // currently incomplete. We need this for the UI improvements that Limi has
     // suggested. See bug 590649.
     cssInfo.rules.forEach(function(cssrule) {
       rules.push(new RuleView(this.tree, cssrule));
     }, this);
 
+    let unmatchedRulesShown = 0;
+
     let data = {
       rules: rules, 
 
       statusText: function(aStatus) {
         return RuleView.STATUS_NAMES[aStatus];
       },
 
       selectorText: function(aSelector) {
@@ -518,33 +548,45 @@ PropertyView.prototype = {
         return result;
       },
 
       l10n: function(aName)
       {
         return self.tree.l10n(aName);
       },
 
+      shouldShowRule: function(aRule) {
+        let result = true;
+        if (!aRule.sheetAllowed) {
+          result = false;
+        }
+
+        if (result && !showAll && aRule.status == CssLogic.STATUS.UNMATCHED) {
+          unmatchedRulesShown++;
+          result = unmatchedRulesShown <= PropertyView.UNMATCHED_RULE_COUNT;
+        }
+
+        return result;
+      },
+
       str: {}
     };
 
-    if (showAll ||
-        cssInfo.unmatchedRuleCount <= PropertyView.UNMATCHED_RULE_COUNT) {
-      data.rules = cssInfo.rules;
+    data.rules = cssInfo.rules;
+
+    if (showAll || cssInfo.filteredUnmatchedRuleCount <=
+      PropertyView.UNMATCHED_RULE_COUNT) {
       data.showMoreUnmatchedRules = false;
-
     } else {
       data.showMoreUnmatchedRules = true;
 
-      data.rules = cssInfo.rules.slice(0, cssInfo.ruleCount -
-          cssInfo.unmatchedRuleCount + PropertyView.UNMATCHED_RULE_COUNT);
-
       data.str.showMoreUnmatchedRules = this.tree.strings.
-          formatStringFromName("style.rule.showMoreUnmatchedRules",
-          [cssInfo.unmatchedRuleCount - PropertyView.UNMATCHED_RULE_COUNT], 1);
+        formatStringFromName("style.rule.showMoreUnmatchedRules",
+          [cssInfo.filteredUnmatchedRuleCount -
+            PropertyView.UNMATCHED_RULE_COUNT], 1);
 
       data.showMoreUnmatchedRulesClick =
         function showMoreUnmatchedRulesClick(aEvent) {
           aEvent.preventDefault();
           self._populateTemplate(true);
         };
     }
 
@@ -574,18 +616,19 @@ PropertyView.prototype = {
    */
   ruleTitle: function PropertyView_ruleTitle(aElement)
   {
     Scheduler.addJob(function() {
       if (!InspectorUI.isStylePanelOpen) {
         return;
       }
 
-      let ruleCount = this.tree.cssLogic.getCssInfo(this.name).ruleCount;
-      if (ruleCount > 1) {
+      let ruleCount = this.tree.cssLogic.getCssInfo(this.name).
+        filteredRuleCount;
+      if (ruleCount > 0) {
         let stringName = ruleCount > 1 ? "numberOfRulesPlural" :
           "numberOfRulesSingular";
         aElement.textContent = this.tree.strings.
           formatStringFromName("style.property." + stringName, [ruleCount], 1);
 
         this.toggle = aElement.ownerDocument.createElement("span");
         this.toggle.className = "toggle";
         this.toggle.innerHTML = "&#x25B6;";
@@ -754,41 +797,40 @@ Templater.prototype.processNode = functi
           value = self.stripBraces(value);
         } else {
           // Extract the loop variable name from 'NAME in ${ARRAY}'.
           let nameArr = value.split(" in ");
           paramName = nameArr[0].trim();
           value = self.stripBraces(nameArr[1].trim());
         }
         recurse = false;
+        aNode.removeAttribute("foreach");
         try {
           let processMember = function(aMember) {
             let clone = aNode.cloneNode(true);
-            clone.removeAttribute("foreach");
             aNode.parentNode.insertBefore(clone, aNode);
             aData[paramName] = aMember;
             self.processNode(clone, aData);
             delete aData[paramName];
           };
           let reply = self.envEval(value, aData, attrs[i].value);
           if (Array.isArray(reply)) {
             reply.forEach(processMember);
           } else {
             for (let param in reply) {
               if (reply.hasOwnProperty(param)) {
                 processMember(param);
               }
             }
           }
           aNode.parentNode.removeChild(aNode);
+          break; // Do not continue any attribute parsing.
         } catch (ex) {
           self.error("Error with '" + value + "'" + ex);
-          recurse = false;
         }
-        aNode.removeAttribute("foreach");
       } else if (name.substring(0, 2) == "on") {
         // Event registration relies on property doing a bind.
         value = self.stripBraces(value);
         let func = self.property(value, aData);
         if (typeof func != "function") {
           self.error("Expected " + value +
             " to resolve to a function, but got " + typeof func);
         }
diff --git a/browser/base/content/csshtmltree.xhtml b/browser/base/content/csshtmltree.xhtml
--- a/browser/base/content/csshtmltree.xhtml
+++ b/browser/base/content/csshtmltree.xhtml
@@ -1,12 +1,12 @@
 <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN"
   "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
-
-<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
+<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en"
+  xmlns:xul="http://www.mozilla.org/keymaster/gatekeeper/there.is.only.xul">
 <head>
   <!-- The title is localized when the Style panel is open by inspector.js -->
   <title>Style</title>
 
   <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/>
   <style type="text/css">
     body {
       font-family: Lucida Grande, sans-serif;
@@ -48,16 +48,21 @@
     .path li:last-child {
       background: -moz-linear-gradient(top, #FFC, #DD8);
     }
     .path li:last-child:after {
       color: red;
       content: "";
     }
 
+    #sheetList, #sheetList menuitem {
+      font-size: 1em;
+      font-weight: normal;
+    }
+
     .sheet_line input {
       vertical-align: middle;
     }
 
     .sheet_line label {
       cursor: pointer;
     }
 
@@ -186,16 +191,26 @@ To visually debug the templates without 
       <label>${l10n('lookingAtLabel')}</label>
       <ol>
         <li foreach="item in ${pathElements}">
           <a href="#" onclick="${pathClick}" 
               __pathElement="${item.element}">${item.display}</a>
         </li>
       </ol>
     </p>
+    <div class="sheetFilter">
+      <xul:label value="${l10n('displaySheetLabel')}" control="sheetList" />
+      <xul:menulist _id="sheetList" label="${l10n('displaySheetLabel')}"
+        oncommand="${sheetChange}"><xul:menupopup><xul:menuitem
+          foreach="sheet in ${cssLogic.sheets}" value="${sheet.href}"
+          label="${sheet.shortSource}" /><xul:menuitem
+          value="all" label="${l10n('displayAllSheets')}" /><xul:menuitem
+          value="ua" label="${l10n('displayBrowserSheets')}" 
+          /></xul:menupopup></xul:menulist>
+    </div>
   </div>
 
   <!--
   templateFooter sits at the bottom of the window showing options.
   For data it needs an instance of CssHtmlTree.
   -->
   <div id="templateFooter">
     <p class="sheet_line">
@@ -254,17 +269,17 @@ To visually debug the templates without 
         title to use for the given rule.
     showMoreUnmatchedRules: true / false, // show a "more unmatched rules" link
     showMoreUnmatchedRulesClick: function() {}, // click event handler for the 
         "show more unmatched rules"
   }
   The table element is only there so the DOM doesn't muck with the tr
   -->
   <table id="templateRules">
-    <tr foreach="rule in ${rules}">
+    <tr foreach="rule in ${rules}" if="${shouldShowRule(rule)}">
       <td width="100%" class="rule-text status${rule.status}">
         ${selectorText(rule.selector)} &#x2192; ${rule.value}
       </td>
       <td class="rule-link"><a target="_blank" href="view-source:${rule.href}"
           class="link" title="${rule.href}">${rule.source}</a></td>
       <td class="rule-score">
         <span title="${specificityTitle(rule)}">
           ${statusText(rule.status)}:${rule.specificity}
diff --git a/browser/base/content/csslogic.js b/browser/base/content/csslogic.js
--- a/browser/base/content/csslogic.js
+++ b/browser/base/content/csslogic.js
@@ -71,16 +71,23 @@ function CssLogic(aStylePanel, aStrings)
   this.sheets = [];
   // Part of the specificity of a rule is its order in the sheet.
   this._ruleCount = 0;
 
   // The cache of examined CSS properties.
   this._cssInfos = {};
   // The computed styles for the viewedElement.
   this._computedStyle = null;
+
+  // Source filter. Only display properties coming from the given source (web
+  // address). Special values:
+  // - "all": will show properties from all style sheets.
+  // - "ua": will show properties from all style sheets, including the browser
+  // styles (user agent ones).
+  this._sourceFilter = "all";
 };
 
 /**
  * Each rule has a status, the bigger the number, the better placed it is to
  * provide styling information.
  */
 CssLogic.STATUS = {
   BEST: 3,
@@ -161,16 +168,50 @@ CssLogic.prototype = {
       this._computedStyle = win.getComputedStyle(this.viewedElement, "");
     } catch (ex) {
       Services.console.logStringMessage("Warning getComputedStyle errored " +
         "for " + this.viewedElement + ". " + ex);
     }
   },
 
   /**
+   * Get the source filter.
+   * @returns {string} The source filter being used.
+   */
+  get sourceFilter() {
+    return this._sourceFilter;
+  },
+
+  /**
+   * Source filter. Only display properties coming from the given source (web
+   * address). Special values:
+   *  - "all": will show properties from all style sheets.
+   *  - "ua": will show properties from all style sheets, including the browser
+   *  styles (user agent ones).
+   */
+  set sourceFilter(aValue) {
+    this._sourceFilter = aValue;
+
+    // Update the CssInfo objects.
+    for (let property in this._cssInfos) {
+      let cssInfo = this._cssInfos[property];
+      if (cssInfo._rules) {
+        cssInfo._findRules();
+        cssInfo._findElementRules();
+        if (!cssInfo._uaRules && this.sourceFilter == "ua") {
+          cssInfo._findUARules();
+        }
+      }
+      if (cssInfo._matched) {
+        cssInfo._markMatches();
+      }
+    }
+  },
+
+  /**
    * Return a CssInfo data structure for the currently viewed element and
    * the specified CSS property. If there is no currently viewed element we
    * return an empty object.
    * 
    * @param {string} aProperty The CSS property to look for.
    * @return a CssInfo structure for the given property
    */
   getCssInfo: function CssLogic_getCssInfo(aProperty)
@@ -276,26 +317,87 @@ CssLogic.getShortNamePath = function Css
       element: aElement
     });
   }
 
   return reply;
 };
 
 /**
+ * Split a rule.selectorText into an array of simpler selectors, taking into
+ * consideration any nested selectors inside parentheses.
+ *
+ * @param {string} selector The rule.selectorText you want split.
+ * @return {array} the array of simpler selectors found in rule.selectorText.
+ */
+CssLogic.selectorTextSplit = function CssLogic_selectorTextSplit(selector) {
+  if (!selector) {
+    return [];
+  }
+
+  selector = selector.trim();
+  if (!selector) {
+    return [];
+  }
+
+  let nesting = 0;
+  let result = [];
+  let currentSelector = [];
+
+  for (let i = 0, n = selector.length; i < n; i++) {
+    let c = selector.charAt(i);
+    switch (c) {
+      case ",":
+        if (nesting == 0 && currentSelector.length > 0) {
+          let newSelector = currentSelector.join("").trim();
+          if (newSelector) {
+            result.push(newSelector);
+          }
+          currentSelector = [];
+        } else {
+          currentSelector.push(c);
+        }
+        break;
+      case "(":
+        nesting++;
+        currentSelector.push(c);
+        break;
+      case ")":
+        nesting--;
+        currentSelector.push(c);
+        break;
+      default:
+        currentSelector.push(c);
+    }
+  }
+
+  // Add the last selector.
+  if (nesting == 0 && currentSelector.length > 0) {
+    let newSelector = currentSelector.join("").trim();
+    if (newSelector) {
+      result.push(newSelector);
+    }
+  }
+
+  return result;
+};
+
+/**
  * A safe each to access cache of bits of information about a stylesheet.
  *
  * @constructor
  * @param {nsIDOMDocument} aDocument the document we inspect.
  * @param {CSSStyleSheet} aDomSheet reference to a DOM CSSStyleSheet object.
+ * @param {boolean} aSystemSheet tells if the stylesheet is system-provided.
  */
-function CssSheet(aDocument, aDomSheet)
+function CssSheet(aDocument, aDomSheet, aSystemSheet)
 {
   this._document = aDocument;
   this._domSheet = aDomSheet;
+  this.systemSheet = aSystemSheet;
 };
 
 CssSheet.prototype = {
   /**
    * Get a source for a stylesheet, taking into account embedded stylesheets
    * for which we need to use document.defaultView.location.href rather than
    * sheet.href
    *
@@ -397,17 +499,19 @@ function CssInfo(aCssLogic, aProperty)
     this.value = aCssLogic._computedStyle.getPropertyValue(aProperty);
   } else {
     this.value = "";
   }
 
   this._matched = false;
   this._localRules = false;
   this._cssLogic = aCssLogic;
+  this._filteredRuleCount = 0;
   this._unmatchedRuleCount = 0;
+  this._filteredUnmatchedRuleCount = 0;
 };
 
 CssInfo.prototype = {
   /**
    * ruleCount could be accessed through this.rules.length, however that means
    * we need to have sorted them into the correct order, which could take time
    * hence we have a separate accessor for the times when you just want to know
    * how rules there are without getting access to the rule ordering.
@@ -417,82 +521,123 @@ CssInfo.prototype = {
   get ruleCount()
   {
     if (!this._rules) {
       this._findRules();
     }
     if (!this._localRules) {
       this._findElementRules();
     }
+    if (!this._uaRules && this._cssLogic.sourceFilter == "ua") {
+      this._findUARules();
+    }
 
     return this._rules.length;
   },
 
   /**
+   * Retrieve the number of filtered rules.
+   *
+   * ruleCount could be accessed through this.rules.length, however that means
+   * we need to have sorted them into the correct order, which could take time
+   * hence we have a separate accessor for the times when you just want to know
+   * how rules there are without getting access to the rule ordering.
+   *
+   * @return {number} the number of filtered rules.
+   */
+  get filteredRuleCount()
+  {
+    this.ruleCount; // Force an update.
+    return this._filteredRuleCount;
+  },
+
+  /**
    * Retrieve the number of unmatched rules.
    * @return {number} the number of unmatched rules.
    */
   get unmatchedRuleCount()
   {
-    if (!this._rules) {
-      this._findRules();
-    }
-    if (!this._localRules) {
-      this._findElementRules();
-    }
+    this.ruleCount; // Force an update.
     if (!this._matched) {
       this._markMatches();
     }
 
     return this._unmatchedRuleCount;
   },
 
   /**
+   * Retrieve the number of unmatched rules, after filtering.
+   * @return {number} the number of unmatched rules, after filtering.
+   */
+  get filteredUnmatchedRuleCount()
+  {
+    this.unmatchedRuleCount; // Force an update.
+
+    return this._filteredUnmatchedRuleCount;
+  },
+
+  /**
    * Accessor for a list of rules sorted by their applicability to the current
    * element.
    *
    * @return {array}
    */
   get rules()
   {
     if (!this._rules) {
       this._findRules();
     }
     if (!this._localRules) {
       this._findElementRules();
     }
+    if (!this._uaRules && this._cssLogic.sourceFilter == "ua") {
+      this._findUARules();
+    }
 
     if (!this._matched) {
       this._markMatches();
     }
 
     return this._rules;
   },
 
   /**
    * Dig through all the sheets looking for matches for <tt>property</tt> and
    * adding the results to <tt>rules</tt>. This method also finds the properties
    * that are defined inline (element.style).
    */
   _findRules: function CssInfo_findRules()
   {
     this._rules = [];
+    this._filteredRuleCount = 0;
 
-    // Loop over all the stylesheets, rules, and properties
+    let filter = this._cssLogic.sourceFilter;
+
+    // Loop over all of the filtered stylesheets, rules, and properties
     for (let s = 0; s < this._cssLogic.sheets.length; s++) {
       let sheet = this._cssLogic.sheets[s];
+      let sheetAllowed = false;
+      if (filter == "all" || filter == "ua" || filter == sheet.href) {
+        sheetAllowed = true;
+      }
+
       for (let r = 0; r < sheet.domRules.length; r++) {
         let domRule = sheet.domRules[r];
         if (domRule.style) {
           for (let t = 0; t < domRule.style.length; t++) {
             if (domRule.style[t] == this.property) {
-              domRule.selectorText.split(",").forEach(function(aSelector) {
+              let selectors = CssLogic.selectorTextSplit(domRule.selectorText);
+              selectors.forEach(function(aSelector) {
                 let rule = new CssRule(this._cssLogic, domRule, this.property,
                     sheet, aSelector.trim());
+                rule.sheetAllowed = sheetAllowed;
                 this._rules.push(rule);
+                if (sheetAllowed) {
+                  this._filteredRuleCount++;
+                }
               }, this);
             }
           }
         }
       }
     }
   },
 
@@ -501,34 +646,90 @@ CssInfo.prototype = {
    */
   _findElementRules: function CssInfo_findElementRules()
   {
     let element = this._cssLogic.viewedElement;
     if (element.style && element.style.getPropertyValue(this.property) != "") {
       let rule = new CssRule(this._cssLogic, {style: element.style},
         this.property, null, "@element.style");
       this._rules.push(rule);
+      this._filteredRuleCount++;
     }
     this._localRules = true;
   },
 
   /**
+   * Find the user-agent rules (Gecko-provided rules).
+   */
+  _findUARules: function CssInfo_findUARules()
+  {
+    let cssLogic = this._cssLogic;
+    let doc = this._cssLogic.viewedDocument;
+    let element = this._cssLogic.viewedElement;
+    let stylePanel = this._cssLogic.stylePanel;
+    let sheets = [];
+    let sheetsMap = {};
+
+    // Find the system sheets for the highlighted element.
+    let matchedRules = stylePanel.domUtils.getCSSStyleRules(element);
+    for (let i = 0, n = matchedRules.Count(); i < n; i++) {
+      let domRule = matchedRules.GetElementAt(i);
+      let domSheet = domRule.parentStyleSheet;
+      let href = domSheet.href;
+      if (!(href in sheetsMap) && domRule.style &&
+        stylePanel.isSystemStyleSheet(domSheet)) {
+        sheets.push(new CssSheet(doc, domSheet, true));
+        sheetsMap[href] = true;
+      }
+    }
+
+    // Find all the rules in each system sheet.
+    for (let i = 0, n = sheets.length; i < n; i++) {
+      let sheet = sheets[i];
+      for (let r = 0, rn = sheet.domRules.length; r < rn; r++) {
+        let domRule = sheet.domRules[r];
+        if (!domRule.style) {
+          continue;
+        }
+        for (let t = 0, tn = domRule.style.length; t < tn; t++) {
+          if (domRule.style[t] != this.property) {
+            continue;
+          }
+
+          let selectors = CssLogic.selectorTextSplit(domRule.selectorText);
+          selectors.forEach(function(aSelector) {
+            let rule = new CssRule(this._cssLogic, domRule, this.property,
+              sheet, aSelector.trim(), true);
+            this._rules.push(rule);
+            this._filteredRuleCount++;
+          }, this);
+        }
+      }
+    }
+
+    this._uaRules = true;
+  },
+
+  /**
    * Go through the <tt>rules</tt> member of <tt>cssInfo</tt> and work out
    * if each rule applies to <tt>viewedElement</tt>.
    */
   _markMatches: function CssInfo_markMatches()
   {
     let parents = [];
     let element = this._cssLogic.viewedElement;
 
     while ((element = element.parentNode) &&
       element.nodeType == Node.ELEMENT_NODE) {
       parents.push(element);
     }
 
+    this._unmatchedRuleCount = 0;
+    this._filteredUnmatchedRuleCount = 0;
+
     for (let rule, r = 0, n = this._rules.length; r < n; r++) {
       rule = this._rules[r];
 
       // Do we have a selector match?
       if (rule.selector == "@element.style" ||
         this._cssLogic.viewedElement.mozMatchesSelector(rule.selector)) {
         rule.status = CssLogic.STATUS.MATCHED;
       } else {
@@ -538,16 +739,19 @@ CssInfo.prototype = {
             rule.status = CssLogic.STATUS.PARENT_MATCH;
             break;
           }
         }
 
         if (rule.status == CssLogic.STATUS.UNKNOWN) {
           rule.status = CssLogic.STATUS.UNMATCHED;
           this._unmatchedRuleCount++;
+          if (rule.sheetAllowed) {
+            this._filteredUnmatchedRuleCount++;
+          }
         }
       }
     }
 
     // Sort the arrays by how well they match the given element.
     this._rules.sort(function(aRule1, aRule2) {
       if (aRule1.status > aRule2.status) {
         return -1;
@@ -573,30 +777,32 @@ CssInfo.prototype = {
    */
   unmarkMatches: function CssInfo_unmarkMatches()
   {
     if (this._rules) {
       let removeRules = [];
 
       this._rules.forEach(function(aRule, aIndex) {
         aRule.status = CssLogic.STATUS.UNKNOWN;
-        if (aRule.selector == "@element.style") {
+        if (aRule.selector == "@element.style" || aRule.systemSheet) {
           removeRules.push(aIndex);
         }
       }, this);
 
       removeRules.forEach(function (aIndex) {
         this._rules.splice(aIndex, 1);
       }, this);
 
       this._unmatchedRuleCount = 0;
+      this._filteredUnmatchedRuleCount = 0;
     }
 
     this._matched = false;
     this._localRules = false;
+    this._uaRules = false;
   },
 
   toString: function CssInfo_toString()
   {
     return "CssInfo[" + this.property + "]";
   },
 };
 
@@ -611,32 +817,36 @@ CssInfo.prototype = {
  * selector and the property. If the rule comes from element.style, then provide
  * an object like this {style: element.style}.
  * @param {string} aProperty the CSS property name we are gathering information
  * for.
  * @param {CSSStyleSheet|null} aSheet reference to the stylesheet where the
  * property and rule were found. Use null if the rule comes from element.style.
  * @param {string} aSelector the simple selector rule. If the rule comes from
  * element.style use "@element.style".
+ * @param {boolean} aSystemSheet tells if the DOMCSSRule comes from a
+ * system-provided stylesheet.
  */
-function CssRule(aCssLogic, aDomRule, aProperty, aSheet, aSelector)
+function CssRule(aCssLogic, aDomRule, aProperty, aSheet, aSelector, aSystemSheet)
 {
   this.property = aProperty;
   this.selector = aSelector;
   this.value = aDomRule.style.getPropertyValue(aProperty);
   let pri = aDomRule.style.getPropertyPriority(aProperty);
 
   let ruleCount;
   if (this.selector == "@element.style") {
     ruleCount = 0;
   } else {
     ruleCount = aCssLogic._ruleCount++;
   }
 
-  this.specificity = new CssSpecificity(this.selector, ruleCount, pri);
+  this.systemSheet = aSystemSheet;
+  this.specificity = new CssSpecificity(this.selector, ruleCount, pri,
+    aSystemSheet);
 
   // domUtils.getRuleLine fails for @import rules.
   this.line = -1;
   if (aCssLogic.stylePanel && this.selector != "@element.style") {
     try {
       this.line = aCssLogic.stylePanel.domUtils.getRuleLine(aDomRule);
     } catch (ex) {
       Services.console.logStringMessage("domUtils.getRuleLine failed for " +
@@ -651,16 +861,17 @@ function CssRule(aCssLogic, aDomRule, aP
   } else {
     this.source = (this.line == -1) ?
         aSheet.shortSource :
         aSheet.shortSource + ":" + this.line;
     this.href = aSheet.href;
   }
 
   this.status = CssLogic.STATUS.UNKNOWN;
+  this.sheetAllowed = true;
 };
 
 CssRule.prototype = {
   toString: function CssInfo_toString()
   {
     return "CssRule[" + this.selector + "->" + this.value + "]";
   },
 };
@@ -670,26 +881,29 @@ CssRule.prototype = {
  * rules.
  *
  * @constructor
  * @param {string} aSelector The selector that we wish to investigate
  * @param {number} aIndex The index of the rule in the HTML document. Later
  * rules have higher precedence (i.e. they are 'more specific')
  * See http://www.w3.org/TR/CSS2/cascade.html#cascade for details.
  * @param {string} aPriority CSS allows only one priority: 'important'
+ * @param {boolean} aSystemSheet tells if the DOMCSSRule comes from a
+ * system-provided stylesheet.
  * @see http://www.w3.org/TR/CSS2/cascade.html#specificity
  * @see http://www.w3.org/TR/CSS2/selector.html
  */
-function CssSpecificity(aSelector, aIndex, aPriority)
+function CssSpecificity(aSelector, aIndex, aPriority, aSystemSheet)
 {
   this.ids = 0;
   this.classes = 0;
   this.tags = 0;
   this.index = aIndex;
   this.important = (aPriority == "important");
+  this.system = aSystemSheet;
 
   // Split on CSS combinators (section 5.2).
   if (aSelector != "@element.style") {
     this.inline = false;
 
     aSelector.split(/[ >+]/).forEach(function(aSimple) {
       // The regex leaves empty nodes combinators like ' > '
       if (!aSimple) {
@@ -712,42 +926,47 @@ function CssSpecificity(aSelector, aInde
     }, this);
   } else {
     this.inline = true;
   }
 
   this.score = 0;
 
   /* Score prefix:
-  0 nothing
-  1 inline
-  2 important
-  3 inline important
+  0 UA normal property
+  1 UA important property
+  2 normal property
+  3 inline
+  4 important
+  5 inline important
   */
-  let scorePrefix = 0;
+  let scorePrefix = this.system ? 0 : 2;
   if (this.inline) {
     scorePrefix++;
   }
   if (this.important) {
-    scorePrefix += 2;
+    scorePrefix += this.system ? 1 : 2;
   }
 
   this.score = "" + scorePrefix + this.ids + this.classes + this.tags;
 };
 
 CssSpecificity.prototype = {
   /**
    * Compare the current CssSpecificity instance to another instance.
    *
    * @param {CssSpecificity} aThat The CssSpecificity instance to compare
    * ourselves against.
    * @return number -1, 0, 1 depending on how aThat compares with this.
    */
   compareTo: function CssSpecificity_compareTo(aThat)
   {
+    if (this.system && !aThat.system) return 1;
+    if (!this.system && aThat.system) return -1;
+
     if (this.inline && !aThat.inline) {
       if (!this.important && aThat.important) return 1;
       else return -1;
     }
 
     if (!this.inline && aThat.inline) {
       if (this.important && !aThat.important) return -1;
       else return 1;
