diff --git a/toolkit/components/console/hudservice/HUDService.jsm b/toolkit/components/console/hudservice/HUDService.jsm
--- a/toolkit/components/console/hudservice/HUDService.jsm
+++ b/toolkit/components/console/hudservice/HUDService.jsm
@@ -227,17 +227,18 @@ ResponseListener.prototype =
 
     storageStream = new StorageStream(8192, aCount, null);
     binaryOutputStream = new BinaryOutputStream(storageStream.getOutputStream(0));
 
     let data = NetUtil.readInputStreamToString(aInputStream, aCount);
 
     if (HUDService.saveRequestAndResponseBodies &&
         this.receivedData.length < RESPONSE_BODY_LIMIT) {
-      this.receivedData += data;
+      this.receivedData += NetworkHelper.
+                           convertToUnicode(data, aRequest.contentCharset);
     }
 
     binaryOutputStream.writeBytes(data, aCount);
 
     let newInputStream = storageStream.newInputStream(0);
     try {
     this.originalListener.onDataAvailable(aRequest, aContext,
         newInputStream, aOffset, aCount);
@@ -385,20 +386,32 @@ var NetworkHelper =
    *        Text to convert.
    * @param string aCharset
    *        Charset to convert the text to.
    * @returns string
    *          Converted text.
    */
   convertToUnicode: function NH_convertToUnicode(aText, aCharset)
   {
+    if (!aCharset) {
+      return aText;
+    }
+
     let conv = Cc["@mozilla.org/intl/scriptableunicodeconverter"].
                createInstance(Ci.nsIScriptableUnicodeConverter);
-    conv.charset = aCharset || "UTF-8";
-    return conv.ConvertToUnicode(aText);
+    conv.charset = aCharset;
+
+    try {
+      return conv.ConvertToUnicode(aText);
+    }
+    catch (ex) {
+      Cu.reportError("NH_convertToUnicode(aText, '" +
+        aCharset + "') exception: " + ex);
+      return aText;
+    }
   },
 
   /**
    * Reads all available bytes from aStream and converts them to aCharset.
    *
    * @param nsIInputStream aStream
    * @param string aCharset
    * @returns string
@@ -2448,53 +2461,58 @@ HUD_SERVICE.prototype =
               httpActivity = item;
               break;
             }
 
             if (!httpActivity) {
               return;
             }
 
-            let msgObject, updatePanel = false;
+            hudId = httpActivity.hudId;
+            let msgObject = httpActivity.messageObject;
+
+            let updatePanel = false;
             let data, textNode;
             // Store the time information for this activity subtype.
             httpActivity.timing[transCodes[aActivitySubtype]] = aTimestamp;
 
             switch (aActivitySubtype) {
               case activityDistributor.ACTIVITY_SUBTYPE_REQUEST_BODY_SENT:
                 if (!self.saveRequestAndResponseBodies) {
                   httpActivity.request.bodyDiscarded = true;
                   break;
                 }
 
-                let gBrowser = HUDService.currentContext().gBrowser;
-
-                let sentBody = NetworkHelper.readPostTextFromRequest(
-                                aChannel, gBrowser);
+                let gBrowser = msgObject.messageNode.ownerDocument.
+                               defaultView.gBrowser;
+                let HUD = HUDService.hudWeakReferences[hudId].get();
+                let browser = gBrowser.
+                              getBrowserForDocument(HUD.contentDocument);
+
+                let sentBody = NetworkHelper.
+                               readPostTextFromRequest(aChannel, browser);
                 if (!sentBody) {
                   // If the request URL is the same as the current page url, then
                   // we can try to get the posted text from the page directly.
                   // This check is necessary as otherwise the
                   //   NetworkHelper.readPostTextFromPage
                   // function is called for image requests as well but these
                   // are not web pages and as such don't store the posted text
                   // in the cache of the webpage.
-                  if (httpActivity.url == gBrowser.contentWindow.location.href) {
-                    sentBody = NetworkHelper.readPostTextFromPage(gBrowser);
+                  if (httpActivity.url == browser.contentWindow.location.href) {
+                    sentBody = NetworkHelper.readPostTextFromPage(browser);
                   }
                   if (!sentBody) {
                     sentBody = "";
                   }
                 }
                 httpActivity.request.body = sentBody;
                 break;
 
               case activityDistributor.ACTIVITY_SUBTYPE_RESPONSE_HEADER:
-                msgObject = httpActivity.messageObject;
-
                 // aExtraStringData contains the response header. The first line
                 // contains the response status (e.g. HTTP/1.1 200 OK).
                 //
                 // Note: The response header is not saved here. Calling the
                 //       aChannel.visitResponseHeaders at this point sometimes
                 //       causes an NS_ERROR_NOT_AVAILABLE exception. Therefore,
                 //       the response header and response body is stored on the
                 //       httpActivity object within the RL_onStopRequest function.
@@ -2512,19 +2530,16 @@ HUD_SERVICE.prototype =
                 msgObject.messageNode.appendChild(
                   msgObject.textFactory(
                     msgObject.prefix +
                     self.getFormatStr("networkUrlWithStatus", data) + "\n"));
 
                 break;
 
               case activityDistributor.ACTIVITY_SUBTYPE_TRANSACTION_CLOSE:
-                msgObject = httpActivity.messageObject;
-
-
                 let timing = httpActivity.timing;
                 let requestDuration =
                   Math.round((timing.RESPONSE_COMPLETE -
                                 timing.REQUEST_HEADER) / 1000);
 
                 // Remove the textNode from the messageNode and add a new one
                 // that contains the request duration.
                 textNode = msgObject.messageNode.firstChild;
diff --git a/toolkit/components/console/hudservice/tests/browser/Makefile.in b/toolkit/components/console/hudservice/tests/browser/Makefile.in
--- a/toolkit/components/console/hudservice/tests/browser/Makefile.in
+++ b/toolkit/components/console/hudservice/tests/browser/Makefile.in
@@ -93,16 +93,17 @@ _BROWSER_TEST_FILES = \
 	browser_webconsole_bug_583816_tab_focus.js \
 	browser_webconsole_bug_594477_clickable_output.js \
 	browser_webconsole_bug_589162_css_filter.js \
 	browser_webconsole_bug_597103_deactivateHUDForContext_unfocused_window.js \
 	browser_webconsole_bug_595350_multiple_windows_and_tabs.js \
 	browser_webconsole_bug_594497_history_arrow_keys.js \
 	browser_webconsole_bug_588342_document_focus.js \
 	browser_webconsole_bug_595934_message_categories.js \
+	browser_webconsole_bug_600183_charset.js \
 	head.js \
 	$(NULL)
 
 # compartment-disabled
 #	browser_webconsole_bug_593003_iframe_wrong_hud.js \
 
 _BROWSER_TEST_PAGES = \
 	test-console.html \
@@ -126,15 +127,17 @@ _BROWSER_TEST_PAGES = \
 	test-bug-595934-css-loader.html \
 	test-bug-595934-css-loader.css \
 	test-bug-595934-css-loader.css^headers^ \
 	test-bug-595934-dom-html.html \
 	test-bug-595934-imagemap.html \
 	test-bug-595934-html.html \
 	test-bug-595934-malformedxml.xhtml \
 	test-bug-595934-svg.xhtml \
+	test-bug-600183-charset.html \
+	test-bug-600183-charset.html^headers^ \
 	$(NULL)
 
 libs:: $(_BROWSER_TEST_FILES)
 	$(INSTALL) $(foreach f,$^,"$f") $(DEPTH)/_tests/testing/mochitest/browser/$(relativesrcdir)
 
 libs:: $(_BROWSER_TEST_PAGES)
 	$(INSTALL) $(foreach f,$^,"$f") $(DEPTH)/_tests/testing/mochitest/browser/$(relativesrcdir)
diff --git a/toolkit/components/console/hudservice/tests/browser/browser_webconsole_bug_600183_charset.js b/toolkit/components/console/hudservice/tests/browser/browser_webconsole_bug_600183_charset.js
new file mode 100644
--- /dev/null
+++ b/toolkit/components/console/hudservice/tests/browser/browser_webconsole_bug_600183_charset.js
@@ -0,0 +1,55 @@
+/* vim:set ts=2 sw=2 sts=2 et: */
+/* ***** BEGIN LICENSE BLOCK *****
+ * Any copyright is dedicated to the Public Domain.
+ * http://creativecommons.org/publicdomain/zero/1.0/
+ *
+ * Contributor(s):
+ *  Mihai È˜ucan <mihai.sucan@gmail.com>
+ *
+ * ***** END LICENSE BLOCK ***** */
+
+const TEST_URI = "http://example.com/browser/toolkit/components/console/hudservice/tests/browser/test-bug-600183-charset.html";
+
+let lastFinishedRequest = null;
+
+function requestDoneCallback(aHttpRequest)
+{
+  lastFinishedRequest = aHttpRequest;
+}
+
+function performTest()
+{
+  ok(lastFinishedRequest, "charset test page was loaded and logged");
+
+  let body = lastFinishedRequest.response.body;
+  ok(body, "we have the response body");
+
+  let chars = "\u7684\u95ee\u5019!"; // çš„é—®å€™!
+  isnot(body.indexOf("<p>" + chars + "</p>"), -1,
+    "found the chinese simplified string");
+
+  lastFinishedRequest = null;
+  HUDService.saveRequestAndResponseBodies = false;
+  HUDService.lastFinishedRequestCallback = null;
+  finishTest();
+}
+
+function test()
+{
+  addTab("data:text/html,Web Console - bug 600183 test");
+
+  browser.addEventListener("load", function(aEvent) {
+    browser.removeEventListener(aEvent.type, arguments.callee, true);
+
+    waitForFocus(function() {
+      openConsole();
+
+      HUDService.saveRequestAndResponseBodies = true;
+      HUDService.lastFinishedRequestCallback = requestDoneCallback;
+
+      browser.addEventListener("load", performTest, true);
+      content.location = TEST_URI;
+    }, content);
+  }, true);
+}
+
diff --git a/toolkit/components/console/hudservice/tests/browser/test-bug-600183-charset.html b/toolkit/components/console/hudservice/tests/browser/test-bug-600183-charset.html
new file mode 100644
--- /dev/null
+++ b/toolkit/components/console/hudservice/tests/browser/test-bug-600183-charset.html
@@ -0,0 +1,9 @@
+<!DOCTYPE HTML>
+<html dir="ltr" xml:lang="en-US" lang="en-US"><head>
+    <meta charset="gb2312">
+    <title>Console HTTP test page (chinese)</title>
+  </head>
+  <body>
+    <p>µÄÎÊºò!</p>
+  </body>
+</html>
diff --git a/toolkit/components/console/hudservice/tests/browser/test-bug-600183-charset.html^headers^ b/toolkit/components/console/hudservice/tests/browser/test-bug-600183-charset.html^headers^
new file mode 100644
--- /dev/null
+++ b/toolkit/components/console/hudservice/tests/browser/test-bug-600183-charset.html^headers^
@@ -0,0 +1,1 @@
+Content-Type: text/html; charset=gb2312
