diff -r 1c7a77f65f9b toolkit/components/console/hudservice/HUDService.jsm
--- a/toolkit/components/console/hudservice/HUDService.jsm	Fri Jul 16 12:55:35 2010 -0300
+++ b/toolkit/components/console/hudservice/HUDService.jsm	Fri Jul 16 21:22:45 2010 +0300
@@ -1,9 +1,10 @@
 /* -*- Mode: js2; js2-basic-offset: 2; indent-tabs-mode: nil; -*- */
+/* vim: set ft=javascript ts=2 et sw=2 tw=80: */
 /* ***** BEGIN LICENSE BLOCK *****
  * Version: MPL 1.1/GPL 2.0/LGPL 2.1
  *
  * The contents of this file are subject to the Mozilla Public License Version
  * 1.1 (the "License"); you may not use this file except in compliance with
  * the License. You may obtain a copy of the License at
  * http://www.mozilla.org/MPL/
  *
@@ -1990,30 +1991,31 @@ function HUDConsole(aHeadsUpDisplay)
     // check to see if logging is on for this level before logging!
     var filterState = HUDService.getFilterState(hudId, aLevel);
 
     if (!filterState) {
       // Ignoring log message
       return;
     }
 
-    let ts = ConsoleUtils.timeStamp(new Date());
+    let ts = ConsoleUtils.timestamp();
     let messageNode = hud.makeHTMLNode("div");
 
     let klass = "hud-msg-node hud-" + aLevel;
 
     messageNode.setAttribute("class", klass);
 
     let argumentArray = [];
     for (var i = 0; i < aArguments.length; i++) {
       argumentArray.push(aArguments[i]);
     }
 
     let message = argumentArray.join(' ');
-    let timestampedMessage = ts + ": " + message;
+    let timestampedMessage = ConsoleUtils.timestampString(ts) + ": " +
+      message;
 
     messageNode.appendChild(chromeDocument.createTextNode(timestampedMessage));
 
     // need a constructor here to properly set all attrs
     let messageObject = {
       logLevel: aLevel,
       hudId: hud.hudId,
       message: message,
@@ -2523,18 +2525,19 @@ LogMessage.prototype = {
    * create a console log div node
    *
    * @returns nsIDOMNode
    */
   createLogNode: function LM_createLogNode()
   {
     this.messageNode = this.elementFactory("div");
 
-    var ts = this.timestamp();
-    var timestampedMessage = ts + ": "  + this.message.message;
+    var ts = ConsoleUtils.timestamp();
+    var timestampedMessage = ConsoleUtils.timestampString(ts) + ": " +
+      this.message.message;
     var messageTxtNode = this.textFactory(timestampedMessage);
 
     this.messageNode.appendChild(messageTxtNode);
 
     var klass = "hud-msg-node hud-" + this.level;
     this.messageNode.setAttribute("class", klass);
 
     var self = this;
@@ -2544,38 +2547,16 @@ LogMessage.prototype = {
       message: self.message,
       timestamp: ts,
       activity: self.activityObject,
       origin: self.origin,
       hudId: self.message.hudId,
     };
 
     this.messageObject = messageObject;
-  },
-
-  timestamp: function LM_timestamp()
-  {
-    // TODO: L10N see bug 568656
-    // TODO: DUPLICATED CODE to be consolidated with the utils timestamping
-    // see bug 568657
-    function logDateString(d)
-    {
-      function pad(n, mil)
-      {
-        if (mil) {
-          return n < 100 ? '0' + n : n;
-        }
-        return n < 10 ? '0' + n : n;
-      }
-      return pad(d.getHours())+':'
-        + pad(d.getMinutes())+':'
-        + pad(d.getSeconds()) + ":"
-        + pad(d.getMilliseconds(), true);
-      }
-    return logDateString(new Date());
   }
 };
 
 
 /**
  * Firefox-specific Application Hooks.
  * Each Gecko-based application will need an object like this in
  * order to use the Heads Up Display
@@ -2638,36 +2619,51 @@ FirefoxApplicationHooks.prototype = {
 /**
  * ConsoleUtils: a collection of globally used functions
  *
  */
 
 ConsoleUtils = {
 
   /**
-   * Generates a millisecond resolution timestamp for console messages
+   * Generates a millisecond resolution timestamp.
    *
-   * @returns string
+   * @returns integer
    */
-  timeStamp: function ConsoleUtils_timeStamp()
+  timestamp: function ConsoleUtils_timestamp()
   {
-    function logDateString(d){
-      function pad(n, mil){
-        if (mil) {
-          return n < 100 ? '0'+n : n;
-        }
-        return n < 10 ? '0'+n : n;
+    return Date.now();
+  },
+
+  /**
+   * Generates a formatted timestamp string for displaying in console messages.
+   *
+   * @param integer [ms] Optional, allows you to specify the timestamp in 
+   * milliseconds since the UNIX epoch.
+   * @returns string The timestamp formatted for display.
+   */
+  timestampString: function ConsoleUtils_timestampString(ms)
+  {
+    // TODO: L10N see bug 568656
+    var d = new Date(ms ? ms : null);
+
+    function pad(n, mil)
+    {
+      if (mil) {
+        return n < 100 ? "0" + n : n;
       }
-      return pad(d.getHours())+':'
-        + pad(d.getMinutes())+':'
-        + pad(d.getSeconds()) + ":"
-        + pad(d.getMilliseconds(), true);
+      else {
+        return n < 10 ? "0" + n : n;
+      }
     }
-    return logDateString(new Date());
-
+
+    return pad(d.getHours()) + ":"
+      + pad(d.getMinutes()) + ":"
+      + pad(d.getSeconds()) + ":"
+      + pad(d.getMilliseconds(), true);
   },
 
   /**
    * Hides a log message by changing its class
    *
    * @param nsIDOMNode aMessageNode
    * @returns nsIDOMNode
    */
