diff --git a/toolkit/components/console/hudservice/tests/browser/Makefile.in b/toolkit/components/console/hudservice/tests/browser/Makefile.in
--- a/toolkit/components/console/hudservice/tests/browser/Makefile.in
+++ b/toolkit/components/console/hudservice/tests/browser/Makefile.in
@@ -82,26 +82,26 @@ _BROWSER_TEST_FILES = \
 	browser_webconsole_bug_585237_line_limit.js \
 	browser_webconsole_bug_581231_close_button.js \
 	browser_webconsole_bug_582201_duplicate_errors.js \
 	browser_webconsole_bug_580454_timestamp_l10n.js \
 	browser_webconsole_netlogging.js \
 	browser_webconsole_bug_594477_clickable_output.js \
 	browser_webconsole_bug_589162_css_filter.js \
 	browser_webconsole_bug_597103_deactivateHUDForContext_unfocused_window.js \
+	browser_webconsole_bug_588342_document_focus.js \
 	head.js \
 	$(NULL)
 
 # compartment-disabled
 #	browser_webconsole_consoleonpage.js \
 #	browser_webconsole_bug_593003_iframe_wrong_hud.js \
 #	browser_webconsole_bug_583816_tab_focus.js \
 #	browser_webconsole_property_provider.js \
 #	browser_webconsole_completion.js \
-# browser_webconsole_bug_588342_document_focus.js \
 
 _BROWSER_TEST_PAGES = \
 	test-console.html \
 	test-network.html \
 	test-network-request.html \
 	test-mutation.html \
 	testscript.js \
 	test-filter.html \
diff --git a/toolkit/components/console/hudservice/tests/browser/browser_webconsole_bug_588342_document_focus.js b/toolkit/components/console/hudservice/tests/browser/browser_webconsole_bug_588342_document_focus.js
--- a/toolkit/components/console/hudservice/tests/browser/browser_webconsole_bug_588342_document_focus.js
+++ b/toolkit/components/console/hudservice/tests/browser/browser_webconsole_bug_588342_document_focus.js
@@ -3,17 +3,17 @@
  * Any copyright is dedicated to the Public Domain.
  * http://creativecommons.org/publicdomain/zero/1.0/
  *
  * Contributor(s):
  *  Mihai Șucan <mihai.sucan@gmail.com>
  *
  * ***** END LICENSE BLOCK ***** */
 
-const TEST_URI = "http://example.com/browser/toolkit/components/console/hudservice/tests/browser/test-console.html";
+const TEST_URI = "data:text/html,Web Console test for bug 588342";
 let fm, notificationBox, input;
 
 function test()
 {
   fm = Cc["@mozilla.org/focus-manager;1"].getService(Ci.nsIFocusManager);
   addTab(TEST_URI);
   browser.addEventListener("load", tabLoad, true);
 }
@@ -52,21 +52,19 @@ function runTest() {
   isnot(fm.focusedWindow, content, "content document has no focus");
 
   let DOMNodeRemoved = false;
   function domNodeRemoved(aEvent) {
     executeSoon(function() {
       if (!DOMNodeRemoved && !notificationBox.querySelector(".hud-box")) {
         DOMNodeRemoved = true;
         document.removeEventListener(aEvent.type, domNodeRemoved, false);
-        is(fm.focusedWindow, browser.contentWindow.wrappedJSObject,
+        is(fm.focusedWindow, browser.contentWindow,
            "content document has focus");
         input = notificationBox = fm = null;
+        finishTest();
       }
-      finishTest();
     });
   }
-  document.addEventListener("DOMNodeRemoved", domNodeRemoved , false);
-  fm = notificationBox = input = null;
-  finishTest();
+  document.addEventListener("DOMNodeRemoved", domNodeRemoved, false);
+  HUDService.deactivateHUDForContext(tab);
 }
 
-
